File: backend/vendoriq-backend/app/models/contract.py
import enum
import uuid
from datetime import datetime, date

from sqlalchemy import String, DateTime, Date, Enum, Float, Text, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class ContractStatus(str, enum.Enum):
    DRAFT = "DRAFT"
    ACTIVE = "ACTIVE"
    EXPIRING_SOON = "EXPIRING_SOON"
    EXPIRED = "EXPIRED"
    TERMINATED = "TERMINATED"
    RENEWED = "RENEWED"


class ComplianceStatus(str, enum.Enum):
    COMPLIANT = "COMPLIANT"
    NON_COMPLIANT = "NON_COMPLIANT"
    UNDER_REVIEW = "UNDER_REVIEW"


class Contract(Base):
    __tablename__ = "contracts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    vendor_id: Mapped[str] = mapped_column(String(36), ForeignKey("vendors.id"), nullable=False)

    contract_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    document_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date] = mapped_column(Date, nullable=False)
    contract_value: Mapped[float | None] = mapped_column(Float, nullable=True)

    status: Mapped[ContractStatus] = mapped_column(Enum(ContractStatus), default=ContractStatus.DRAFT)
    compliance_status: Mapped[ComplianceStatus] = mapped_column(Enum(ComplianceStatus), default=ComplianceStatus.UNDER_REVIEW)

    auto_renew: Mapped[bool] = mapped_column(default=False)
    terms: Mapped[str | None] = mapped_column(Text, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_by: Mapped[str | None] = mapped_column(String(36), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    vendor = relationship("Vendor", back_populates="contracts")
    certifications = relationship("Certification", back_populates="contract", cascade="all, delete-orphan")


class Certification(Base):
    __tablename__ = "certifications"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    contract_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("contracts.id"), nullable=True)
    vendor_id: Mapped[str] = mapped_column(String(36), ForeignKey("vendors.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    issuing_authority: Mapped[str | None] = mapped_column(String(200), nullable=True)
    issue_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    expiry_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    document_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    contract = relationship("Contract", back_populates="certifications")

File: frontend/src/models/contract.ts
/**
 * Module 6: Contract & Compliance Management
 * Data structures for contracts, certifications, compliance records,
 * vendor documents, renewals, and contract notifications.
 *
 * Frontend-only mock structures. Swapping in real HTTP/REST API calls
 * requires editing only contractService.ts — these interfaces remain unchanged.
 */

// ── Contract ─────────────────────────────────────────────────────────────────

export type ContractStatus = 'Draft' | 'Active' | 'Expired' | 'Renewed' | 'Terminated';
export type ContractType =
  | 'Fixed Price'
  | 'Time & Materials'
  | 'Service Agreement'
  | 'Supply Agreement'
  | 'AMC'
  | 'NDA'
  | 'Maintenance Contract'
  | 'License Agreement'
  | 'Framework Contract'
  | 'SLA Agreement';

export interface Contract {
  contractId: number;
  contractNumber: string;        // format: CT-YYYY-XXXX
  contractTitle: string;
  vendorId: number;
  vendorName: string;
  contractType: ContractType;
  procurementCategory: string;
  startDate: string;             // ISO date string
  endDate: string;               // ISO date string
  contractValue: number;         // in INR
  paymentTerms: string;
  sla: string;                   // SLA description
  warrantyDetails: string;
  responsibleManager: string;
  status: ContractStatus;
  documentName: string;
  linkedProcurementNumbers: string[];
  createdAt: string;             // ISO datetime string
  daysToExpiry: number;          // computed: negative = already expired
}

// ── Contract Renewal ──────────────────────────────────────────────────────────

export interface ContractRenewal {
  renewalId: number;
  contractId: number;
  contractNumber: string;
  vendorName: string;
  oldEndDate: string;
  newEndDate: string;
  renewedBy: string;
  renewalDate: string;
  remarks: string;
}

// ── Certification ─────────────────────────────────────────────────────────────

export type CertificationStatus = 'Valid' | 'Expiring Soon' | 'Expired';

export interface Certification {
  certificationId: number;
  vendorId: number;
  vendorName: string;
  certificationName: string;
  certificateNumber: string;
  issuingAuthority: string;
  issueDate: string;
  expiryDate: string;
  documentName: string;
  status: CertificationStatus;
  daysToExpiry: number;          // computed: negative = already expired
}

// ── Compliance Record ─────────────────────────────────────────────────────────

export type ComplianceStatus =
  | 'Compliant'
  | 'Pending Verification'
  | 'Non-Compliant'
  | 'Expired';

export interface ComplianceRecord {
  complianceId: number;
  vendorId: number;
  vendorName: string;
  requirementName: string;       // e.g. 'GST Registration', 'ISO Quality Standards'
  complianceStatus: ComplianceStatus;
  lastVerifiedDate: string;
  verifiedBy: string;
  remarks: string;
}

// ── Vendor Document ───────────────────────────────────────────────────────────

export type VendorDocumentType =
  | 'GST Certificate'
  | 'PAN Card'
  | 'Company Registration'
  | 'Business License'
  | 'Bank Details'
  | 'Insurance'
  | 'Product Catalog'
  | 'NDA'
  | 'Service Agreement'
  | 'Quality Certificate';

export interface VendorDocument {
  documentId: number;
  vendorId: number;
  vendorName: string;
  documentType: VendorDocumentType;
  documentName: string;
  uploadedBy: string;
  uploadedAt: string;            // ISO datetime string
  version: string;               // e.g. 'v1.0', 'v2.3'
}

// ── Contract Notification ─────────────────────────────────────────────────────

export type ContractNotificationType =
  | 'Contract Expiry'
  | 'Certification Expiry'
  | 'Compliance Alert';

export type NotificationSeverity = 'Info' | 'Warning' | 'Critical';

export interface ContractNotification {
  notificationId: number;
  type: ContractNotificationType;
  /** Contract number or certificate number */
  referenceNumber: string;
  vendorName: string;
  expiryDate: string;
  remainingDays: number;         // negative = already expired
  renewalStatus: string;         // e.g. 'Renewal Pending', 'Auto-Flagged', 'Renewed'
  severity: NotificationSeverity;
  createdAt: string;             // ISO datetime string
  read: boolean;
}

// ── Service Filter/Pagination types ──────────────────────────────────────────

export interface ContractFilters {
  status?: ContractStatus | 'All';
  contractType?: ContractType | 'All';
  vendorId?: number;
  search?: string;
  page?: number;
  pageSize?: number;
}

export interface CertificationFilters {
  status?: CertificationStatus | 'All';
  vendorId?: number;
  search?: string;
}

export interface DocumentFilters {
  vendorId?: number;
  documentType?: VendorDocumentType | 'All';
  search?: string;
}

export interface PaginatedContracts {
  items: Contract[];
  total: number;
  page: number;
  pageSize: number;
}

export interface ContractSummaryMetrics {
  totalContracts: number;
  activeContracts: number;
  expiringContracts: number;
  expiredContracts: number;
  totalCertifications: number;
  expiredCertifications: number;
  compliancePercentage: number;
}

